sql/logictest: TestLogic failed

[(sql/logictest).TestLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2075591&tab=buildLog) on [master@58e91fc5923b34ccda4c143d0a22753e071859d2](https://github.com/cockroachdb/cockroach/commits/58e91fc5923b34ccda4c143d0a22753e071859d2):

```
=== RUN   TestLogic/fakedist-disk/cascade/DeleteSetDefault_ToUpdateCascade
--- progress: testdata/logic_test/alter_table: 294 statements/queries
--- done: testdata/logic_test/distsql_srfs with config fakedist-disk: 7 tests, 0 failures

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

In the logictest artifacts I see:
```
panic: flow: unexpected 10240 leftover bytes

goroutine 9445361 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic(0x4bea800, 0xc04acc6540, 0xc05239ea80, 0x4207721, 0x20, 0xc00f164db8, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:335 +0x23c
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc02d3210e0, 0x4bea800, 0xc04acc6540, 0x163f901)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:397 +0x2c3
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:384
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3206
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0xc007ac8fc0, 0x4bea800, 0xc04acc6540)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:422 +0x78
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Cleanup(0xc011388990, 0x4bea800, 0xc04acc6540)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:309 +0x1ea
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow.func2(0x4c62220, 0xc011388990, 0x4bea800, 0xc04acc6540)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:104 +0x59
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:102 +0x27d
```

[(sql/logictest).TestLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2084318&tab=buildLog) on [master@4c2da99f879bc2bc3922bc421ac8fee80995b134](https://github.com/cockroachdb/cockroach/commits/4c2da99f879bc2bc3922bc421ac8fee80995b134):

```
=== RUN   TestLogic
    TestLogic: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic179030437
    TestLogic: test_log_scope.go:58: use -show-logs to present logs inline
--- FAIL: TestLogic (960.18s)
=== RUN   TestLogic/local/txn_stats
=== PAUSE TestLogic/local/txn_stats
=== CONT  TestLogic/local/txn_stats
        --- FAIL: TestLogic/local/txn_stats (1.50s)
=== RUN   TestLogic/local
    --- FAIL: TestLogic/local (0.55s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -470,6 +470,7 @@ func (r opResult) createAndWrapRowSource(
 	// own, so the used memory will be accounted for.
 	r.Op, r.IsStreaming = c, true
 	r.MetadataSources = append(r.MetadataSources, c)
+	r.ToClose = append(r.ToClose, c)
 	return nil
 }
 

--- a/pkg/sql/colexec/columnarizer.go
+++ b/pkg/sql/colexec/columnarizer.go
@@ -138,8 +138,11 @@ func (c *Columnarizer) Run(context.Context) {
 	colexecerror.InternalError("Columnarizer should not be Run")
 }
 
-var _ colexecbase.Operator = &Columnarizer{}
-var _ execinfrapb.MetadataSource = &Columnarizer{}
+var (
+	_ colexecbase.Operator       = &Columnarizer{}
+	_ execinfrapb.MetadataSource = &Columnarizer{}
+	_ Closer                     = &Columnarizer{}
+)
 
 // DrainMeta is part of the MetadataSource interface.
 func (c *Columnarizer) DrainMeta(ctx context.Context) []execinfrapb.ProducerMetadata {
@@ -152,6 +155,12 @@ func (c *Columnarizer) DrainMeta(ctx context.Context) []execinfrapb.ProducerMeta
 		c.accumulatedMeta = append(c.accumulatedMeta, *meta)
 	}
 	return c.accumulatedMeta
+}
+
+// Close is part of the Operator interface.
+func (c *Columnarizer) Close(ctx context.Context) error {
+	c.input.ConsumerClosed()
+	return nil
 }
 
 // ChildCount is part of the Operator interface.

--- a/pkg/sql/colexec/columnarizer_test.go
+++ b/pkg/sql/colexec/columnarizer_test.go
@@ -100,6 +100,10 @@ func TestColumnarizerDrainsAndClosesInput(t *testing.T) {
 	require.True(t, len(meta) == 1)
 	require.True(t, testutils.IsError(meta[0].Err, errMsg))
 	require.True(t, rb.Done)
+	require.Equal(t, execinfra.DrainRequested, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
+	// Closing the Columnarizer should call ConsumerClosed on the processor.
+	require.NoError(t, c.Close(ctx))
+	require.Equal(t, execinfra.ConsumerClosed, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
 }
 
 func BenchmarkColumnarize(b *testing.B) {

